### PR TITLE
feat: add metadata API

### PR DESCRIPTION
**Pull request by Sh4d1**
Signed-off-by: Patrik Cyvoct <patrik@ptrk.io>

### CODE REVIEW

File: api/metadata/v1/metadata_sdk.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+// NewAPI returns a API object from a Scaleway client.
+func NewAPI() *API {
+	return &API{}
+}
+
+// Metadata returns the metadata avialable from the server
+func (*API) Metadata() (*Metadata, error) {
+	resp, err := http.Get(metadataURL + "/conf?format=json")
+	if err != nil {
+		return nil, err
```
Comment (kindermoumoute) [inline api/metadata/v1/metadata_sdk.go]: should return a [SdkError](https://github.com/scaleway/scaleway-sdk-go/blob/master/scw/errors.go#L8)

File: api/metadata/v1/metadata_sdk.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	metadata := &Metadata{}
+	err = json.NewDecoder(resp.Body).Decode(metadata)
+	if err != nil {
+		return nil, err
```
Comment (kindermoumoute) [inline api/metadata/v1/metadata_sdk.go]: should return a [SdkError](https://github.com/scaleway/scaleway-sdk-go/blob/master/scw/errors.go#L8)

File: api/metadata/v1/metadata_sdk.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+// API metadata API
+type API struct {
+}
+
+// NewAPI returns a API object from a Scaleway client.
+func NewAPI() *API {
+	return &API{}
+}
+
+// Metadata returns the metadata avialable from the server
```
Comment (kindermoumoute) [inline api/metadata/v1/metadata_sdk.go]: GetMetadata

File: api/metadata/v1/metadata_sdk.go (hunk 1 above)
Comment (Sh4d1) [reply api/metadata/v1/metadata_sdk.go]: fixed

File: api/metadata/v1/metadata_sdk.go (hunk 3 above)
Comment (Sh4d1) [reply api/metadata/v1/metadata_sdk.go]: fixed

File: api/metadata/v1/metadata_sdk.go (hunk 2 above)
Comment (Sh4d1) [reply api/metadata/v1/metadata_sdk.go]: fixed

File: api/metadata/v1/metadata_sdk.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+		}
+		_, err = userdataClient.Do(request)
+		if err != nil {
+			retries++ // retry with a different source port
+			continue
+		}
+
+		return nil
+	}
+	return errors.New("too may bind port retries for DeleteUserdata")
```
Comment (DenBeke) [inline api/metadata/v1/metadata_sdk.go]: typo: `may` -> `many`

Comment (Sh4d1) [reply api/metadata/v1/metadata_sdk.go]: fixed

File: api/metadata/v1/metadata_sdk.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+	if key == "" {
+		return errors.New("key must not be empty in SetUserdata")
+	}
+
+	retries := 0
+	for retries <= metadataRetryBindPort {
+		port := rand.Intn(1024)
+		localTCPAddr, err := net.ResolveTCPAddr("tcp", ":"+strconv.Itoa(port))
+		if err != nil {
+			return errors.New(err.Error())
```
Comment (jerome-quere) [inline api/metadata/v1/metadata_sdk.go]: use errors.Wrap

Comment (Sh4d1) [reply api/metadata/v1/metadata_sdk.go]: fixed

**Pull request merged**